### PR TITLE
Add operations for manipulating meta data

### PR DESCRIPTION
**Pull request by PicoCentauri**
This PR adds an operations to manipulate Labels as well as the Labels columns stored in TensorMaps as `keys`, `samples` or `properties`.

----
:books: Documentation preview :books:: https://equistore--299.org.readthedocs.build/en/299/

### CODE REVIEW

**Comment by PicoCentauri** [comment]
Sorry for the many force pushes. The three functions should be there now.

File: python/equistore-operations/equistore/operations/manipulate_metadata.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+def _remove_from_list(a: List, b: List):
+    """Remove all items in ``b`` from ``a``"""
+    for item in b:
+        if item not in a:
+            raise ValueError(f"'{item}' not in list")
+        a.remove(item)
+
+
+def append_metadata(
+    tensor: TensorMap, name: str, values: Union[float, np.ndarray], axis: str
```
Comment (PicoCentauri) [inline python/equistore-operations/equistore/operations/manipulate_metadata.py:53]: I currently only allow for only one new label name. We can extend this to allow also a list of names. However, this will makes verifying the values more complex...

Comment (jwa7) [reply python/equistore-operations/equistore/operations/manipulate_metadata.py:53]: Yes I think a single name is fine for now

**Review by jwa7 — CHANGES_REQUESTED** [review]
Thanks for this @PicoCentauri ! 

3 things on this:

1. I think the name of the functions would be better as specifying explicitly that we are adding/removing columns. 
2. As a user I would like to have the option to insert the column at an index of my choosing. Or, at the very least, have also the option to prepend as well as append (though controlling the index is preferred).
3. From an UX-consistency point of view I would order the args `axis, name, values`. It fits better with other operations that use the `axis, values` or `axis, labels` ordering.

Can I therefore suggest a function with a signature like:
```py
def insert_metadata_column(
    tensor: TensorMap, axis: str, name: str, values: Union[float, np.ndarray], col_idx: Optional[int] = None
) -> TensorMap:
```
where for the default behaviour with `col_idx=None` the column is appended to the end, i.e. as is currently done by this function.

Thoughts?

File: python/equistore-operations/equistore/operations/__init__.py [hunk 2]
```diff
@@ -44,6 +44,7 @@
 )
 from .join import join
 from .lstsq import lstsq
+from .manipulate_metadata import append_metadata, rename_metadata
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/__init__.py]: Hmm But then we can not use them as toplevel imports because calling `equistore.append` is not clear...

**Comment by PicoCentauri** [comment]
I renamed the functions and introduced the Labels methods that I needed.

Let me know what you think.

File: python/equistore-core/equistore/core/labels.py [hunk 3]
```diff
@@ -303,6 +303,141 @@ def range(name: str, end: int) -> "Labels":
             values=np.arange(end, dtype=np.int32).reshape(-1, 1),
         )
 
+    def append(self, name: str, values: np.ndarray) -> "Labels":
+        """Append a new column to the end.
+
+        :param name: name of the new column
+        :param values: 1D array of values for the new column
```
Comment (PicoCentauri) [inline python/equistore-core/equistore/core/labels.py:310]: I demand this to be a 1D array because we are stacking a new column to a 2D array. If you think this seems inconsistent w.r.t. to creating a new Labels object (where we demand a 2D array as input).

File: python/equistore-core/equistore/core/labels.py [hunk 4]
```diff
@@ -303,6 +303,144 @@ def range(name: str, end: int) -> "Labels":
             values=np.arange(end, dtype=np.int32).reshape(-1, 1),
         )
 
+    def append(self, name: str, values: np.ndarray) -> "Labels":
```
Comment (Luthaf) [inline python/equistore-core/equistore/core/labels.py:306]: If we add function to the Labels in equistore-core, we need to add the same ones in equistore-torch. 

The API is nicer if we add member functions, but that's more work to do. I don't have strong feelings between adding functions twice or keeping them in equistore-operations

Comment (PicoCentauri) [reply python/equistore-core/equistore/core/labels.py:306]: Yes, I will add them to `equistore-torch` if we agree on the API. I think the fit quite good here. I only haven't touched this to avoid changing lines twice...

File: python/equistore-operations/equistore/operations/manipulate_labels_column.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    >>> equistore.remove_labels_column(tensor, axis="keys", name="extra")
+    TensorMap with 1 blocks
+    keys: keys
+           0
+    """
+    _check_axis(axis)
+
+    keys = tensor.keys
+    if axis == "keys":
+        keys = keys.remove(name=name)
```
Comment (jwa7) [inline python/equistore-operations/equistore/operations/manipulate_labels_column.py:204]: If we're left with an empty Labels object, an error is created when using this function but the message is not helpful. I'd suggest we check for this and give a better error message:

```suggestion
        keys = keys.remove(name=name)
        if len(keys) == 0:
            raise ValueError("Cannot remove the only key name present.")
```

File: python/equistore-operations/equistore/operations/manipulate_labels_column.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+    blocks = []
+    for block in tensor:
+        samples = block.samples
+        properties = block.properties
+
+        if axis == "samples":
+            samples = samples.remove(name)
```
Comment (jwa7) [inline python/equistore-operations/equistore/operations/manipulate_labels_column.py:212]: And maybe here too with the samples and properties
```suggestion
            samples = samples.remove(name)
            if len(samples) == 0:
                raise ValueError("Cannot remove the only samples name present.")
```

File: python/equistore-operations/equistore/operations/manipulate_labels_column.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        elif axis == "properties":
+            properties = properties.remove(name)
```
Comment (jwa7) [inline python/equistore-operations/equistore/operations/manipulate_labels_column.py:214]:
```suggestion
            properties = properties.remove(name)
            if len(properties) == 0:
                raise ValueError("Cannot remove the only properties name present.")
```

File: python/equistore-operations/tests/manipulate_labels_column.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+    for block in new_tensor:
+        assert_equal(block.properties.values[:, -1], values)
+
+        for _, gradient in block.gradients():
+            assert gradient.properties.names[-1] == "foo"
+            assert_equal(gradient.properties.values[:, -1], values)
+
+
+def test_append_unknonw_axis():
```
Comment (jwa7) [inline python/equistore-operations/tests/manipulate_labels_column.py:82]:
```suggestion
def test_append_unknown_axis():
```

File: python/equistore-operations/tests/manipulate_labels_column.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+    for block in new_tensor:
+        assert_equal(block.properties.values[:, index], values)
+
+        for _, gradient in block.gradients():
+            assert gradient.properties.names[index] == "foo"
+            assert_equal(gradient.properties.values[:, index], np.arange(80))
+
+
+def test_insert_unknonw_axis():
```
Comment (jwa7) [inline python/equistore-operations/tests/manipulate_labels_column.py:140]:
```suggestion
def test_insert_unknown_axis():
```

File: python/equistore-operations/tests/manipulate_labels_column.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        new="foo",
+    )
+    assert new_tensor.property_names[0] == "foo"
+
+    for block in new_tensor:
+        for _, gradient in block.gradients():
+            assert gradient.properties.names[0] == "foo"
+
+
+def test_rename_unknonw_axis():
```
Comment (jwa7) [inline python/equistore-operations/tests/manipulate_labels_column.py:187]:
```suggestion
def test_rename_unknown_axis():
```

File: python/equistore-operations/tests/manipulate_labels_column.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        name="extra",
+    )
+    assert new_tensor.property_names == ["properties"]
+
+    for block in new_tensor:
+        for _, gradient in block.gradients():
+            assert gradient.properties.names == ["properties"]
+
+
+def test_remove_unknonw_axis():
```
Comment (jwa7) [inline python/equistore-operations/tests/manipulate_labels_column.py:226]:
```suggestion
def test_remove_unknown_axis():
```

File: python/equistore-operations/tests/manipulate_labels_column.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+    for block in new_tensor:
+        for parameter, gradient in block.gradients():
+            if parameter == "positions":
+                gradient_sample_names = ["sample", "foo", "atom"]
+            elif parameter == "cell":
+                gradient_sample_names = ["sample"]
+            assert gradient.samples.names == gradient_sample_names
+
+
+def test_renames_properties():
```
Comment (jwa7) [inline python/equistore-operations/tests/manipulate_labels_column.py:171]:
```suggestion
def test_rename_properties():
```

**Review by jwa7 — COMMENTED** [review]
Super super useful thanks @PicoCentauri !!

**Comment by PicoCentauri** [comment]
Thanks for the review @jwa7 ! I put the length check in the beginning of the function. We don't have to do this for every block.

I will now do the port of the Labels method for equsitore-torch.

File: equistore-torch/include/equistore/torch/labels.hpp [hunk 13]
```diff
@@ -79,6 +79,18 @@ class EQUISTORE_TORCH_EXPORT LabelsHolder: public torch::CustomClassHolder {
         return values_;
     }
 
+    /// Append a new dimension to the end
```
Comment (Luthaf) [inline equistore-torch/include/equistore/torch/labels.hpp:82]:
```suggestion
    /// Append a new dimension with the given `name` and `values`  to the end of these `Labels`
```

File: equistore-torch/include/equistore/torch/labels.hpp [hunk 14]
```diff
@@ -79,6 +79,18 @@ class EQUISTORE_TORCH_EXPORT LabelsHolder: public torch::CustomClassHolder {
...
+    TorchLabels append(std::string name, torch::Tensor values);
+
+    /// Insert a new dimension before index
```
Comment (Luthaf) [inline equistore-torch/include/equistore/torch/labels.hpp:85]:
```suggestion
    /// Insert a new dimension with the given `name` and `values` before `index` in these `Labels`
```

File: equistore-torch/include/equistore/torch/labels.hpp [hunk 15]
```diff
@@ -79,6 +79,18 @@ class EQUISTORE_TORCH_EXPORT LabelsHolder: public torch::CustomClassHolder {
...
+    TorchLabels insert(int64_t index, std::string name, torch::Tensor values);
+
+    /// Remove `name` from the dimensions of these Labels
```
Comment (Luthaf) [inline equistore-torch/include/equistore/torch/labels.hpp:88]:
```suggestion
    /// Remove `name` from the dimensions of these `Labels`
```

File: equistore-torch/include/equistore/torch/labels.hpp [hunk 16]
```diff
@@ -79,6 +79,18 @@ class EQUISTORE_TORCH_EXPORT LabelsHolder: public torch::CustomClassHolder {
...
+    TorchLabels remove(std::string name);
+
+    /// Remove `name` from the dimensions of these Labels
```
Comment (Luthaf) [inline equistore-torch/include/equistore/torch/labels.hpp:91]:
```suggestion
    /// Rename `old_name` dimensions in these Labels to `new_name`
```

File: equistore-torch/src/labels.cpp [hunk 17]
```diff
@@ -265,6 +265,73 @@ LabelsHolder LabelsHolder::to_owned() const {
...
+            ValueError,
+            "`values` must be a 1D tensor"
+        );
+    }
+
+    auto old_values = this->values();
+    auto first = old_values.index({torch::indexing::Slice(), torch::indexing::Slice(0, index)});
+    auto second = old_values.index({torch::indexing::Slice(), torch::indexing::Slice(index)});
+
+    auto new_values = torch::hstack({first, values.reshape({1, static_cast<int64_t>(values.size(0))}), second});
```
Comment (Luthaf) [inline equistore-torch/src/labels.cpp:290]:
```suggestion
    auto new_values = torch::hstack({first, values.reshape({values.size(0), 1}), second});
```

`values.size(0)` should already be a 64-bit int. Also, I think the shape was the wrong way around?

File: python/equistore-core/equistore/core/labels.py [hunk 18]
```diff
@@ -436,6 +436,144 @@ def values(self) -> np.ndarray:
         """
         return self._values
 
+    def append(self, name: str, values: np.ndarray) -> "Labels":
+        """Append a new dimension to the end.
```
Comment (Luthaf) [inline python/equistore-core/equistore/core/labels.py:440]:
```suggestion
        """Append a new dimension to the end of the :py:class:`Labels`.
```

File: python/equistore-core/equistore/core/labels.py [hunk 19]
```diff
@@ -436,6 +436,144 @@ def values(self) -> np.ndarray:
...
+        >>> label.append(name="bar", values=np.array([10]))
+        Labels(
+            foo  bar
+            42   10
+        )
+        """
+        return self.insert(index=len(self), name=name, values=values)
+
+    def insert(self, index: int, name: str, values: np.ndarray) -> "Labels":
+        """Insert a new dimension before ``index``.
```
Comment (Luthaf) [inline python/equistore-core/equistore/core/labels.py:462]:
```suggestion
        """Insert a new dimension before ``index`` in the :py:class:`Labels`.
```

File: python/equistore-core/equistore/core/labels.py [hunk 20]
```diff
@@ -436,6 +436,144 @@ def values(self) -> np.ndarray:
...
+
+        if len(values.shape) != 1:
+            raise ValueError("`values` must be a 1D array")
+
+        new_values = np.insert(self.values, index, values, axis=1)
+
+        return Labels(names=new_names, values=new_values)
+
+    def remove(self, name: str) -> "Labels":
+        """Remove ``name`` from the dimensions of these Labels.
```
Comment (Luthaf) [inline python/equistore-core/equistore/core/labels.py:496]:
```suggestion
        """Remove ``name`` from the dimensions of the :py:class:`Labels`.
```

File: python/equistore-core/equistore/core/labels.py [hunk 21]
```diff
@@ -436,6 +436,144 @@ def values(self) -> np.ndarray:
...
+        new_names = self.names.copy()
+        new_names.remove(name)
+
+        index = self.names.index(name)
+        new_values = np.delete(self.values, index, axis=1)
+
+        return Labels(names=new_names, values=new_values)
+
+    def rename(self, old: str, new: str) -> "Labels":
+        """Rename the ``old`` dimension to ``new``.
```
Comment (Luthaf) [inline python/equistore-core/equistore/core/labels.py:547]:
```suggestion
        """Rename the ``old`` dimension to ``new`` in the :py:class:`Labels`.
```

File: python/equistore-core/tests/labels.py [hunk 22]
```diff
@@ -111,6 +111,44 @@ def test_custom_constructors():
         Labels.range(0, 1)
 
 
+def test_custom_manipulation():
```
Comment (Luthaf) [inline python/equistore-core/tests/labels.py:114]:
```suggestion
def test_dimensions_manipulation():
```

File: python/equistore-operations/equistore/operations/__init__.py [hunk 23]
```diff
@@ -44,6 +44,12 @@
 )
 from .join import join
 from .lstsq import lstsq
+from .manipulate_labels_column import (
+    append_labels_column,
+    remove_labels_column,
+    rename_labels_column,
+    insert_labels_column,
```
Comment (Luthaf) [inline python/equistore-operations/equistore/operations/__init__.py:51]: Naming-wise do we want `labels_column` or `_dimension`? Both are fine for me, with a slight preference for `_dimension`.


```suggestion
    append_dimension,
    remove_dimension,
    rename_dimension,
    insert_dimension,
```

File: python/equistore-torch/equistore/torch/documentation.py [hunk 24]
```diff
@@ -298,6 +298,111 @@ def __ne__(self, other: "Labels") -> bool:
         different values)
         """
 
+    def append(self, name: str, values: torch.Tensor) -> "Labels":
+        """Append a new dimension to the end.
```
Comment (Luthaf) [inline python/equistore-torch/equistore/torch/documentation.py:302]: Same suggestions here as in equistore-core

File: python/equistore-torch/tests/labels.py [hunk 25]
```diff
@@ -384,6 +384,41 @@ def test_intersection():
     assert torch.all(second_mapping == torch.LongTensor([-1, 0, -1]))
 
 
+def test_manipulation():
```
Comment (Luthaf) [inline python/equistore-torch/tests/labels.py:387]:
```suggestion
def test_dimensions_manipulation():
```

File: python/equistore-torch/tests/labels.py [hunk 26]
```diff
@@ -384,6 +384,41 @@ def test_intersection():
...
+    label = Labels("foo", torch.tensor([[42]]))
+
+    # # Labels.insert
```
Comment (Luthaf) [inline python/equistore-torch/tests/labels.py:390]:
```suggestion
    # Labels.insert
```

**Review by Luthaf — COMMENTED** [review]
Look good! The main discussion point is about the operations naming!

File: python/equistore-operations/equistore/operations/__init__.py (hunk 23 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/__init__.py:51]: I also like the `_dimensions` a bit more. @jwa7  seems to be in favor of columns?

File: equistore-torch/src/labels.cpp (hunk 17 above)
Comment (PicoCentauri) [reply equistore-torch/src/labels.cpp:290]: yes it is the other way around. But why is this also working...

File: python/equistore-operations/equistore/operations/__init__.py (hunk 23 above)
Comment (Luthaf) [reply python/equistore-operations/equistore/operations/__init__.py:51]: these functions operate on TensorMap, the dimensions of which are always stored in Labels (each name/column in the Labels is a dimension of the high-dimensional sparse array we represent with a TensorMap). So for me the `_labels` part is redundant =)

File: equistore-torch/src/labels.cpp (hunk 17 above)
Comment (Luthaf) [reply equistore-torch/src/labels.cpp]: not sure, maybe some torch magic/dispatch?

File: python/equistore-operations/equistore/operations/__init__.py (hunk 23 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/__init__.py:51]: Fine for me, lets just wait if @jwa7 is also fine.

Comment (jwa7) [reply python/equistore-operations/equistore/operations/__init__.py:51]: @PicoCentauri @Luthaf Yep sounds good to me!

**Comment by PicoCentauri** [comment]
I did the renaming. If you two @Luthaf and @jwa7 are happy we can merge this.

**Review by jwa7 — APPROVED** [review]
Looks good, thanks @PicoCentauri ! :)

**Comment by PicoCentauri** [comment]
@Luthaf if you are also happy you can merge this.

**Comment by Luthaf** [comment]
Looks good!
